### PR TITLE
Expose `UnknownBlock` error via `ApiError`

### PR DESCRIPTION
**Pull request by tdimitrov**
In [certain cases](https://github.com/paritytech/polkadot/issues/5885) a runtime api is called for an unknown block. For example a block which is already pruned or on an abandon fork.

In such cases the correct error is returned but it is wrapped in `ApiError::Application` and the only way to figure out what is the problem is to inspect the actual message in the error. In polkadot for example this usually happens when the runtime api version is being queried. It's beneficial to be able to clearly separate such errors so i that when they occur the client side can handle them more gracefully. E.g. log less stressful error message than `State already discarded for BlockId` or cancel any pending work related on this block.

A potential problem I've just realised is `ApiError` exhaustive enum so this change might break a lot of code.

### CODE REVIEW

**Comment by tdimitrov** [comment]
This is still work in progress. I hadn't got a chance to test it on versi due to more important tests being executed there.

**Comment by tdimitrov** [comment]
This change seems to have the job done:
image
I'm looking for some feedback now :)

File: primitives/api/src/lib.rs [hunk 1]
```diff
@@ -519,6 +519,8 @@ pub enum ApiError {
 	StateBackendIsNotTrie,
 	#[error(transparent)]
 	Application(#[from] Box<dyn std::error::Error + Send + Sync>),
+	#[error("Api called for an unknownBlock: {0}")]
```
Comment (bkchr) [inline primitives/api/src/lib.rs:522]:
```suggestion
	#[error("Api called for an unknown Block: {0}")]
```

**Review by bkchr — APPROVED** [review]
Not sure there is that much too say :P The variant is relative generic, but this is already because the underlying error is not giving you better information.

**Review by michalkucharczyk — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2285462

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2285463

**Review by koute — APPROVED** [review]

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2285767

**Pull request merged**